### PR TITLE
Added the possibility to cancel goals and fixes the preempt/ execution timeouts

### PR DESCRIPTION
**Pull request by RichardvdK**
Sorry for the "big diff", but this file needed some love. :)
Implement canceling of goals as that was not possible before

The biggest changes are:
- Fixed the execution and preemption timeout timers
- Made 'outcomes' a param to make it compatible with your state machine
- Updated state names to ROS2 convention (preempt -> cancel)
- added extra intermediate internal state for preemption and used state enum instead of global vars
- added function to wait for cancelation request response with timeout

### CODE REVIEW

**Comment by MCFurry** [comment]
LGTM! I think we can easily point this to the forked repo?

**Comment by Timple** [comment]
I don't see anything surprising, thanks for picking this up.
I'd forward this PR to the parent repository.

File: smach_ros/smach_ros/simple_action_state.py [hunk 1]
```diff
@@ -208,76 +214,108 @@ def __init__(self,
...
+                if self.node.get_clock().now() - self._activate_time > self._exec_timeout:
+                    self.node.get_logger().warn(
+                        f"Action {self._action_name} timed out after {self._exec_timeout.nanoseconds / 1e9} seconds. "
+                        f"Trying to cancel goal with timeout {self._preempt_timeout.nanoseconds / 1e9}.")
+                    self._status = self.__cancel_goal_and_wait(
+                        timeout=self._preempt_timeout)
+                    with self._done_cond:
+                        self._done_cond.notify()
+                time.sleep(0.1)
+            except Exception as e:
```
Comment (Rayman) [inline smach_ros/smach_ros/simple_action_state.py]: Can you catch a more specific exception here?

File: smach_ros/smach_ros/simple_action_state.py [hunk 2]
```diff
@@ -377,56 +430,95 @@ def execute(self, ud):
...
+                'STATUS_ABORTED')
             if i < len(strs):
                 return strs[i]
             else:
-                return 'UNKNOWN ('+str(i)+')'
+                return 'UNKNOWN (' + str(i) + ')'
 
-        gh = future.result()
+        if future.done():
+            goal_result = future.result()
```
Comment (Rayman) [inline smach_ros/smach_ros/simple_action_state.py]: I think this if statement is not needed

File: smach_ros/smach_ros/simple_action_state.py (hunk 1 above)
Comment (RichardvdK) [reply smach_ros/smach_ros/simple_action_state.py]: Not really... there are quite some possible exceptions to catch here

File: smach_ros/smach_ros/simple_action_state.py (hunk 2 above)
Comment (RichardvdK) [reply smach_ros/smach_ros/simple_action_state.py]: Agree!

File: smach_ros/smach_ros/simple_action_state.py [hunk 3]
```diff
@@ -208,76 +214,108 @@ def __init__(self,
...
+        # If status is active, try to cancel goal with a timeout
+        if self._status == ActionState.ACTIVE and rclpy.ok():
+            self.node.get_logger().info(
+                f"Preempt on action '{self._action_name}' trying to cancel goal: {self._client_goal_handle} "
+                f"with timeout {self._preempt_timeout.nanoseconds / 1e9}")
+            self._status = self.__cancel_goal_and_wait(self._preempt_timeout)
+            with self._done_cond:
+                self._done_cond.notify()
+
+    def __cancel_goal_and_wait(self, timeout) -> ActionState:
```
Comment (Timple) [inline smach_ros/smach_ros/simple_action_state.py]: Private functions should start with a single underscore, not two.

File: smach_ros/smach_ros/simple_action_state.py (hunk 1 above)
Comment (Timple) [reply smach_ros/smach_ros/simple_action_state.py]: I agree with @Rayman here. Catch only expected exceptions. Catching all might even be undesired at some levels.
If you don't know all, start with the one you encountered and we can expand the list later.

**Pull request merged**